### ISSUE TITLE
Streaming support from HTTP outbound endpoint

### ISSUE THREAD

**Issue by asarkar** [spring-cloud/spring-cloud-stream#1152]
I'm looking to stream a large file (~20GB) from a HTTP GET endpoint to database records. The file is gzipped so I'd have to convert it to a zip stream first, and then read the content. Obviously, I'm trying to avoid writing to disk and unzipping, otherwise I wouldn't be asking. Is there anything in the framework that supports this use case? Before going down to parsing bytes myself, I thought this is not such an uncommon scenario, and there might be something existing. So far, I've found nothing. `org.springframework.core.io.buffer.DataBufferUtils` comes close, but it appears to close the stream before downstream consumers have consumed fully (not sure about that, didn't dig a whole lot).

**Comment by artembilan** [comment]
Looks like you need this one:

```
/**
 * {@code HttpMessageReader} for parsing {@code "multipart/form-data"} requests
 * to a stream of {@link Part}'s using the Synchronoss NIO Multipart library.
 *
 * <p>This reader can be provided to {@link MultipartHttpMessageReader} in order
 * to aggregate all parts into a Map.
 *
 * @author Sebastien Deleuze
 * @author Rossen Stoyanchev
 * @author Arjen Poutsma
 * @since 5.0
 * @see <a href="https://github.com/synchronoss/nio-multipart">Synchronoss NIO Multipart</a>
 * @see MultipartHttpMessageReader
 */
public class SynchronossPartHttpMessageReader implements HttpMessageReader<Part> {
```

https://github.com/sdeleuze/webflux-multipart

**Comment by asarkar** [comment]
@artembilan, thanks, I'll take a look. I wrote the following meanwhile using OKHttp. They seem to do the unzip automatically based on the content type, so I don't have to.

```
public Message<Flux<String>> doTransform(Message<String> message) throws Exception {
    String url = message.getPayload();
    String mimeType = (String) message.getHeaders().getOrDefault(FILE_TYPE_HEADER, ALL_VALUE);

    Request.Builder builder = new Request.Builder()
            .url(url);

    headers
            .forEach((String key, Object value) -> builder.header(key, value.toString()));
    Request request = builder
            .header(ACCEPT, mimeType)
            .build();

    List<Closeable> closeables = new ArrayList<>();

    Flux<String> flux = Mono.<Response>create(emitter -> {
        try {
            Response response = httpClient
                    .newCall(request)
                    .execute();
            closeables.add(response);
            int statusCode = response.code();

            if (HttpStatus.resolve(statusCode).is2xxSuccessful()) {
                emitter.success(response);
            } else {
                emitter.error(new MessagingException(message, String.format("Failed to download from URL: %s, " +
                        "response status code: %d.", url, statusCode)));
            }
        } catch (IOException e) {
            emitter.error(e);
        }
    })
            .flatMapMany(response -> {
                InputStream body = response.body().byteStream();
                InputStream is = null;
                Reader reader = null;
                BufferedReader br = null;
                closeables.addAll(asList(br, reader, is, body));

                try {
                    is = getGZIPStreamWithFallback(body);
                    reader = new InputStreamReader(is, StandardCharsets.UTF_8);
                    br = new BufferedReader(reader, bufferSize);

                    return Flux.fromStream(br.lines())
                            .doOnError(t -> log.error("Failed to download from URL: {}, response status code: {}.",
                                    url, response.code(), t));
                } catch (IOException e) {
                    return Flux.error(e);
                }
            })
            .doAfterTerminate(() -> {
                log.info("Closing response stream.");
                closeQuietly(closeables);
            });

    MessageBuilder<Flux<String>> messageBuilder = MessageBuilder
            .withPayload(flux);

    LinkedCaseInsensitiveMap<Object> headers = new LinkedCaseInsensitiveMap<>();
    headers.putAll(message.getHeaders());
    messageBuilder.copyHeaders(headers);

    return messageBuilder.build();
}
```

**Comment by artembilan** [comment]
Yeah... I got you. Sorry for confuse.
And looks like this your JIRA is definitely related: https://jira.spring.io/browse/SPR-15948

**Comment by artembilan** [comment]
OK. I think I'm messing everything here, sorry. I give up. Not sure how to help you.

Although, give us, please, a clue how is it related to [Spring Cloud Stream](https://cloud.spring.io/spring-cloud-stream)?

**Comment by asarkar** [comment]
> how is it related to Spring Cloud Stream?

Isn't Spring Cloud Stream all about streaming data? There are components to stream files from places like S3, why not HTTP?

**Comment by artembilan** [comment]
Well, this is streaming, indeed, but as a distributed messaging flow in the cloud. There is nothing about low-level `InputStream` job. 

Since you didn't find an out-of-the-box solution, then there is no one yet. However I guess we still can write a simple `FluxHttpMessageConverter` to convert `inputMessage.getBody()` to the `Flux<>`. Similar to what we have with the `ByteArrayHttpMessageConverter` and inject it into the `httpclient` Stream App...

Would you mind to express your strict requirement here? What would you like to see as a fix from the Framework perspective? Otherwise let's close it as non-related.

Thanks

**Comment by asarkar** [comment]
If this feature is implemented in the framework, I'd like to see support for the following:

Ability to get a `Flux<DataBuffer>` or `Flux<InputStream>` from WebFlux Outbound gateway; even better, with support for automatic unzipping based on relevant headers like `Content-Type` and `Content-Encoding` (OkHttp does this). Probably related to https://github.com/spring-projects/spring-integration/issues/2300.

**Comment by artembilan** [comment]
Well, sorry, but that's really not related to Spring Cloud Stream.

Closing this one in favor of that in Spring Integration, although, IMO. that's more Spring WebFlux `WebClient` request, than Spring Integration's Channel Adapter, but we'll see when I come to that SI issue. Maybe tomorrow...

Somehow Splitter/Aggregator is a bit different request and should be handled in different ticket. And again: outside of Spring Cloud Stream, too.

Does it sound like a plan? 
I got your requirements, but I'm not sure how to help from the SCSt perspective.
